### PR TITLE
Fix n2f.f90.

### PR DESCRIPTION
**Pull request by schweitzpgi**
Rewrite filter.
Add another raiseBaseType method.

### CODE REVIEW

File: flang/lib/Lower/ConvertExpr.cpp [hunk 1]
```diff
@@ -2404,6 +2404,19 @@ class ArrayExprLowering {
       return result;
     }
 
+    /// In an explicit space, the context may include an implicit subspace. The
+    /// RHS of the assignment does not necessarily have rank and can be promoted
+    /// from a scalar to an array. In that case, the implicit subscripts must be
+    /// removed.
```
Comment (jeanPerier) [inline flang/lib/Lower/ConvertExpr.cpp:2410]: As I understand the comment, removeImplicit() is called for RHS like:
```
type t
  real :: x(100)
end type
real :: y(100)
type(t) :: a(100)
Forall (i = 1:100)
  a(i)%x = y(i)
end forall 
```

Is my understanding correct here ?

What about the case where the RHS is a mix of scalar and array reference (e.g. `a(i)%x = y(i) + z`, with z an array) ? What is ensuring that `removeImplicit` is called for `y(i)` but not `z` ?

File: flang/lib/Lower/ConvertExpr.cpp [hunk 2]
```diff
@@ -3212,17 +3238,21 @@ class ArrayExprLowering {
...
-      operator()(const Fortran::evaluate::ArrayRef &ref) {
+      RT operator()(const Fortran::evaluate::ArrayRef &ref) {
         if ((ref.base().IsSymbol() || ref.base().Rank() == 0) &&
             ref.Rank() > 0 && !ref.subscript().empty()) {
+          auto baseTy = ael->raiseBaseType(ref.base());
+          auto ty = ref.base().IsSymbol()
+                        ? baseTy
+                        : baseTy.cast<fir::RecordType>().getType(
+                              toStringRef(ref.GetLastSymbol().name()));
+          auto bounds = ael->getShape(ty);
```
Comment (jeanPerier) [inline flang/lib/Lower/ConvertExpr.cpp:3255]: Does this assume that `ty` shape is compile time constant ?

**Review by jeanPerier — APPROVED** [review]

File: flang/lib/Lower/ConvertExpr.cpp (hunk 1 above)
Comment (schweitzpgi) [reply flang/lib/Lower/ConvertExpr.cpp:2410]: Good point. Fixed this with another commit.

File: flang/lib/Lower/ConvertExpr.cpp (hunk 2 above)
Comment (schweitzpgi) [reply flang/lib/Lower/ConvertExpr.cpp:3252]: Yes. This bag of worms is still a TODO.

**Pull request merged**